server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -550,12 +551,19 @@ final SearchContext createAndPutContext(ShardSearchRequest request) throws IOExc
         SearchContext context = createContext(request);
         boolean success = false;
         try {
-            putContext(context);
             if (request.scroll() != null) {
                 openScrollContexts.incrementAndGet();
                 context.indexShard().getSearchOperationListener().onNewScrollContext(context);
             }
             context.indexShard().getSearchOperationListener().onNewContext(context);
+            putContext(context);
```

Comment:
I don't understand why you moved the `putContext(context)`

Reply:
By registering the search context in `activeContexts` after having invoked `onNewContext/onNewScrollContext`, we guarantee that for a specific `SearchContext`, the call to `onNewXXX` happens before the matching call the `onFreeXXX`.

Reply:
I see I didn't think about `SearchService.afterIndexRemoved`.  my issue here is that we have to call  `onFreeXXX` but if one of the `onNewContext /onNewScrollContext` fails we don't register and fail? I think we need extra protection for this?